Breakpoints.

Keindel opened the pull request:
Added pretty json representation of Body in Breaks.

In Break add/edit dialog String Field has been splitted into multiple fields: Protocol, Host, Path, Query.

Add request and/or response choice in breaks: choice of break activation on-request and/or on-response is now available.

Add break persisting. Breaks in UI are now consistent with session DB.

Neglect escaping of url special chars for breaks UI.

kingthorin commented:
Also I believe there's a JSON View available via the Marketplace.

kingthorin commented:
To address the DCO requirement you'll need to sign-off the commit(s):

- https://github.com/zaproxy/zaproxy/blob/main/CONTRIBUTING.md#developer-certificate-of-origin
- https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff

thc202 commented on build.gradle.kts:
```diff
@@ -33,6 +33,8 @@ allprojects {
     project.plugins.withType(JavaPlugin::class) {
         dependencies {
             "errorprone"("com.google.errorprone:error_prone_core:2.18.0")
+            "implementation"("com.fasterxml.jackson.core:jackson-core:2.14.2")
+            "implementation"("com.fasterxml.jackson.core:jackson-databind:2.14.2")
```
It shouldn't be added anywhere in core.

Keindel commented:
Json formatting is excluded, DCO requirement is satisfied.

thc202 commented on zap/src/main/java/org/zaproxy/zap/db/sql/SqlDatabase.java:
```diff
@@ -229,6 +230,12 @@ public TableStructure getTableStructure() {
 
+    @Override
+    public TableBreak getTableBreak() {
+        // NOT IMPLEMENTED YET
+        return null;
```
What happens then?

thc202 commented on zap/src/main/java/org/zaproxy/zap/extension/brk/ExtensionBreak.java:
```diff
@@ -211,19 +215,103 @@ private BreakpointsPanel getBreakpointsPanel() {
+    }
+
+    private void writeBreakToDB(BreakpointMessageInterface breakpoint) {
+        HttpBreakpointMessage httpBreakpoint = (HttpBreakpointMessage) breakpoint;
```
What happens with other breakpoint types? e.g. WebSocket

Keindel replied:
I got confused at first. HttpBreakpointMessage is the only implementation in core.
Extensions should have their own implementations to work with DB.
So I add instance class checks before methods call.

kingthorin commented:
- The commits can be squashed or fixed-up, and signed off for DCO.
- The code formatting should be consistent. It is syntactically correct, and I'm sure there's other old code where things weren't done 'right' but conditionals should use curly braces even for single statements, etc.

Keindel commented:
Added null-checks for not yet implemented experimental DB case.
Added instance class checks before DB-related methods call.

New commits are squashed and signed off for DCO.

Keindel commented:
It seems that neglecting of special chars escaping for breaks UI may affect user workflow.
So I've reverted those changes.

kingthorin commented on zap/src/main/java/org/parosproxy/paros/db/paros/ParosDatabase.java:
Should have a ZAP comment in the header/leadin

kingthorin commented on zap/src/main/java/org/zaproxy/zap/db/RecordBreak.java:
```diff
@@ -0,0 +1,118 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2003-2004 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
If this is a new class it should have an up-to-date license block

kingthorin commented on zap/src/main/java/org/zaproxy/zap/db/TableBreak.java:
```diff
@@ -0,0 +1,58 @@
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2015 The ZAP Development Team
```
2023

kingthorin commented on zap/src/main/java/org/zaproxy/zap/db/paros/ParosTableBreak.java:
```diff
@@ -0,0 +1,296 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2003-2004 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
Should have an up-to-date header.

Keindel commented:
updated header comments
